cpp/src/groupby/sort/group_nunique.cu

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/types.hpp>
+#include <cudf/aggregation.hpp>
+
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+
+namespace cudf {
+namespace experimental {
+namespace groupby {
+namespace detail {
+
+
+std::unique_ptr<column> group_nunique(
+    column_view const& values,
+    rmm::device_vector<size_type> const& group_labels,
+    size_type num_groups,
+    rmm::device_vector<size_type> const& group_offsets,
+    include_nulls _include_nulls,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  CUDF_EXPECTS(num_groups >= 0, "number of groups cannot be negative");
+  CUDF_EXPECTS(static_cast<size_t>(values.size()) == group_labels.size(),
+               "Size of values column should be same as that of group labels");
+
+  auto result = make_numeric_column(data_type(type_to_id<size_type>()),
+                  num_groups, mask_state::UNALLOCATED, stream, mr);
+
+  if (num_groups == 0) {
+    return result;
+  }
+
+  auto values_view = column_device_view::create(values);
+  if (values.nullable()) {
```

Comment:
@devavret  other groupby aggregations uses `nullable()` too. Is there any particular reason? can we change it to `has_nulls()`?

Reply:
`has_nulls()` incurs additional kernel call but saves us from using an expensive kernel in case of all valids. I don't think there's going to be any noticeable difference in actual performance.

Reply:
> other groupby aggregations uses nullable() too

Which ones? The only one that does it is `group_count` and it's because it's essentially computing per group the same thing that `has_nulls` kernel does for whole series.

Reply:
I think it should be has_nulls() here.